### PR TITLE
Testing/bookmarks

### PR DESCRIPTION
**Pull request by kspeer825**
# Description of change
Test incremental and full table non-report based streams. Ensure that the bookmarked value is respected for a given incremental stream. Ensure and that full table syncs are consistent.

# Manual QA steps
 - None
 
# Risks
 - None, test additions.
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_bookmarks.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                    for message in first_sync_messages:
+                        replication_key_value = message.get('data').get(replication_key)
+                        self.assertLessEqual(replication_key_value, first_bookmark_value_utc,
+                                             msg="First sync bookmark was set incorrectly, a record with a greater rep key value was synced")
+
+                    # Verify the second sync bookmark value is the max replication key value for a given stream
+                    for message in second_sync_messages:
+                        replication_key_value = message.get('data').get(replication_key)
+                        self.assertLessEqual(replication_key_value, second_bookmark_value_utc,
+                                             msg="Second sync bookmark was set incorrectly, a record with a greater rep key value was synced")
```
Comment (luandy64) [inline tests/test_bookmarks.py:178]: Could this be combined with the previous loop over `second_sync_messages`?

**Review by luandy64 — CHANGES_REQUESTED** [review]
I'm blocking just for my comments to be answered, I think the changes here fine, especially since the tests pass

File: tests/test_bookmarks.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        Formatting of state
+           {'bookmarks': {'accounts': {'last_record': '2020-11-23T03:58:39.610000+00:00'}}}
+        """
+
+        stream_to_current_state = dict()
+        for stream, bookmark in current_state['bookmarks'].items():
+            stream_to_current_state[stream] = bookmark.get(self.get_bookmark_key(stream))
+
+        stream_to_calculated_state = {stream: "" for stream in current_state['bookmarks'].keys()}
```
[the remark this answers is not in the document]
Comment (kspeer825) [reply tests/test_bookmarks.py:61]: Making change.

File: tests/test_bookmarks.py (hunk 1 above)
Comment (kspeer825) [reply tests/test_bookmarks.py:178]: Making change.

File: tests/base.py [hunk 3]
```diff
@@ -763,3 +771,53 @@ def get_all_fields(self):
...
+
+            for bk_value in bk_values:
+                if bk_value is None:
+                    continue
+
+                if max_bookmarks[prefixed_stream][stream_bookmark_key] is None:
+                    max_bookmarks[prefixed_stream][stream_bookmark_key] = bk_value
+
+                if bk_value > max_bookmarks[prefixed_stream][stream_bookmark_key]:
+                    max_bookmarks[prefixed_stream][stream_bookmark_key] = bk_value
```
[the remark this answers is not in the document]
Comment (kspeer825) [reply tests/base.py:821]: That would simplify it. We are doing a str comparison on ISO date-times. I think to be safe we would need to do the start of the epoch or whatever it's called. The minimum value for a date-time.

File: tests/test_bookmarks.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+        for stream, state in stream_to_current_state.items():
+            # convert state from string to datetime object
+            state_as_datetime = dateutil.parser.parse(state)
+            # subtract 1 day from the state
+            calculated_state_as_datetime = state_as_datetime - datetime.timedelta(days=1)
+            # convert back to string and format
+            calculated_state = str(calculated_state_as_datetime).replace(' ', 'T')
+            stream_to_calculated_state[stream] = calculated_state
```
[the remark this answers is not in the document]
Comment (kspeer825) [reply tests/test_bookmarks.py:70]: Making change

File: tests/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kspeer825) [reply tests/base.py:821]: I want to wait to refactor leaving a TODO for now. Next PR is for testing bookmarks for report streams only. I will use that to refactor base further.

**Review by luandy64 — APPROVED** [review]

**Pull request merged**